### DISCUSSION TITLE
Mypy and static type checking

### DISCUSSION THREAD

**Discussion by jlashner** [simonsobs/socs#724] [General]
For a number of reasons I think OCS and SOCS would benefit a lot from proper typing and static type checking. The main reason being that tests currently are very expensive, taking ~30 minutes to run, however the majority of bugs that we catch are issues that could have been found by type-checking with mypy, given that our functions are properly typed.

How do people feel about pushing static typing and encouraging developers to do this? We could use the "agent robustness" push as an opportunity to roll out static typing to individual agents.

One large barrier to full coverage is dynamically generated objects such as OCS Clients and session data, however I do not think this should prevent us from typing other parts of our code.

**Comment by BrianJKoopman** [comment]
I'm not totally against it, but I haven't used mypy before. Let me try it out and get back to you.

This still doesn't fix the underlying problem of tests taking a long time to run. Fundamentally I think our problem is related to the package structure. We're testing every agent all the time, which is why it's so expensive. Often PRs are relevant to a singular agent. If only the tests for that agent ran it would be a lot quicker. I don't quite know how to solve this in a way that's simple to maintain.

**Comment by BrianJKoopman** [comment]
Following up on the discussion today. I'm onboard with running mypy on these repos, starting with `ocs`. We should start with a PR solely focused on setting up a mypy config and getting existing code to run with mypy, i.e. fix any existing errors. Once that's setup we can go ahead and add annotations. Essentially we should follow these recommendations: https://mypy.readthedocs.io/en/stable/existing_code.html

We went through this before when adding `autopep8` and similar. Let's follow that model and add a mypy hook in pre-commit.
